File: 125-ValidPalindrome/125-ValidPalindrome.java
// Last updated: 8/16/2026, 9:52:21 PM
class Solution {
    public boolean isPalindrome(String s) {
        // // replace non-alphanumeric chars to empty
        // String newS = s.replaceAll("[^a-zA-Z0-9]", "").toLowerCase();
        // System.out.println(newS);
        // int n = newS.length();
        // int p1 = 0, p2 = n - 1;
        // while(p1 < n / 2) {
        //     if (newS.charAt(p1) != newS.charAt(p2)) return false;
        //     p1 ++;
        //     p2 --;
        // }
        // return true;

        // or without regex
        int p1 = 0, p2 = s.length() - 1;
        while(p1 <= p2) {
            if ( !Character.isLetterOrDigit(s.charAt(p1)) ) {p1++; continue;}
            if ( !Character.isLetterOrDigit(s.charAt(p2)) ) {p2--; continue;}
            if ( Character.toLowerCase(s.charAt(p1)) != Character.toLowerCase(s.charAt(p2)) ) return false;
            p1 ++;
            p2 --;
        }
        return true;
    }
}

File: 125-ValidPalindrome/125-ValidPalindrome.py
# Last updated: 9/16/2026, 4:07:19 PM
1class Solution:
2    def isPalindrome(self, s: str) -> bool:
3        left, right = 0, len(s) - 1
4
5        while left < right:
6            while left < right and not s[left].isalnum():
7                left += 1
8            while left < right and not s[right].isalnum():
9                right -= 1
10            
11            if s[left].lower() != s[right].lower():
12                return False
13            
14            left += 1
15            right -= 1
16        
17        return True